// @cpdevtools/lib-node-utilities@0.4.54/install/platformInstallers/BashInstaller.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BashInstaller = void 0;
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const path_1 = __importDefault(require("path"));
const utils_1 = require("../../utils");
const PlatformInstallerBase_1 = require("./PlatformInstallerBase");
class BashInstaller extends PlatformInstallerBase_1.PlatformInstallerBase {
    get isInstalled() {
        return (async () => {
            return true;
        })();
    }
    async installOrUpdate() {
        await this.execScript(this.installOrUpdateScript);
    }
    async update() {
        await this.execScript(this.updateScript);
    }
    async uninstall() {
        await this.execScript(this.uninstallScript);
    }
    async execScript(script, cwd) {
        const p = path_1.default.join(__dirname, "__exec_script.sh");
        try {
            if ((0, fs_1.existsSync)(p)) {
                await (0, promises_1.rm)(p, { force: true });
            }
            await (0, promises_1.writeFile)(p, script, { encoding: "utf-8" });
            await (0, utils_1.exec)(`chmod +x ${p}`);
            await (0, utils_1.exec)(p, { cwd });
        }
        finally {
            if ((0, fs_1.existsSync)(p)) {
                await (0, promises_1.rm)(p, { force: true });
            }
        }
    }
}
exports.BashInstaller = BashInstaller;
