// @cpdevtools/lib-node-utilities@0.4.54/install/InstallerService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalInstallerService = exports.InstallerService = void 0;
const dependency_graph_1 = require("dependency-graph");
const fast_glob_1 = __importDefault(require("fast-glob"));
const fs_1 = require("fs");
const linq_1 = __importDefault(require("linq"));
const posix_1 = __importDefault(require("path/posix"));
const dep_graph_1 = require("../utils/dep_graph");
const hooks_1 = require("./hooks");
const AfterInstallOrUpdate_1 = require("./hooks/AfterInstallOrUpdate");
const AfterUninstall_js_1 = require("./hooks/AfterUninstall.js");
const AfterUpdate_js_1 = require("./hooks/AfterUpdate.js");
const BeforeInstallOrUpdate_1 = require("./hooks/BeforeInstallOrUpdate");
const BeforeUninstall_1 = require("./hooks/BeforeUninstall");
const BeforeUpdate_1 = require("./hooks/BeforeUpdate");
const Installer_1 = require("./installer/Installer");
const detectPlatform_1 = require("./platform/detectPlatform");
const Platform_1 = require("./platform/Platform");
class InstallerService {
    constructor() {
        Object.defineProperty(this, "_init", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_scanDirs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "_installers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Map()
        });
        this.scanDir(posix_1.default.join(__dirname, "_installers_"));
    }
    scanDir(path) {
        if (!this._scanDirs.includes(path)) {
            this._scanDirs.push(path);
        }
        return this;
    }
    async scanForInstallers() {
        const files = (await Promise.all(this._scanDirs.filter((d) => (0, fs_1.existsSync)(d)).map((d) => (0, fast_glob_1.default)(["*.installer.js", "*/*.installer.js"].map((p) => posix_1.default.join(d, p)))))).flat();
        const installers = files
            .map((f) => require(f))
            .filter((m) => !!m.default)
            .map((m) => (Array.isArray(m.default) ? m.default : [m.default]))
            .map((m) => {
            return m.filter((i) => (0, Installer_1.isValidInstaller)(i));
        })
            .flat();
        return installers;
    }
    init() {
        if (!this._init) {
            this._init = (async () => {
                const scannedInstallers = await this.scanForInstallers();
                scannedInstallers.forEach((inst) => this._addInstaller(inst));
            })();
        }
        return this._init;
    }
    _addInstaller(installer) {
        if ((0, Installer_1.isValidInstaller)(installer)) {
            this._installers.set(installer.id, installer);
        }
    }
    addInstaller(installer) {
        if (!this._init) {
            this._addInstaller(installer);
        }
        else {
            throw new Error(`Install Service already initialized. Can't add any more installers.`);
        }
    }
    get installers() {
        return (async () => {
            await this.init();
            return linq_1.default.from(this._installers.values());
        })();
    }
    async getInstallerById(id, forPlatform) {
        return (await this.installers).firstOrDefault((i) => i.id === id && (!forPlatform || typeof i.platforms === "function" || !!i.platforms[forPlatform]));
    }
    async getPlatformInstallerById(id, forPlatform) {
        if (!forPlatform) {
            const detected = (0, detectPlatform_1.detectPlatform)();
            if (detected === Platform_1.Platform.UNKNOWN) {
                throw new Error("Could not detect Platform");
            }
            forPlatform = detected;
        }
        const platforms = (await this.getInstallerById(id, forPlatform))?.platforms;
        return typeof platforms === "function" ? platforms : platforms?.[forPlatform];
    }
    async hasInstaller(id, forPlatform) {
        return !!(await this.getInstallerById(id, forPlatform));
    }
    async createInstallerInstance(id) {
        const platform = (0, detectPlatform_1.detectPlatform)();
        if (platform === Platform_1.Platform.UNKNOWN) {
            return undefined;
        }
        const installer = await this.getPlatformInstallerById(id, platform);
        return installer ? new installer() : undefined;
    }
    async isInstalled(id) {
        const inst = await this.createInstallerInstance(id);
        return (await inst?.isInstalled) ?? false;
    }
    async installOrUpdateById(idOrInstaller) {
        const inst = typeof idOrInstaller === "string" ? await this.createInstallerInstance(idOrInstaller) : idOrInstaller;
        const wasInstalled = (await inst?.isInstalled) ?? false;
        if (wasInstalled) {
            if ((0, BeforeUpdate_1.implementsBeforeUpdate)(inst)) {
                if ((await inst.beforeUpdate()) === false) {
                    return;
                }
            }
        }
        else {
            if ((0, hooks_1.implementsBeforeInstall)(inst)) {
                if ((await inst.beforeInstall()) === false) {
                    return;
                }
            }
        }
        if ((0, BeforeInstallOrUpdate_1.implementsBeforeInstallOrUpdate)(inst)) {
            if ((await inst.beforeInstallOrUpdate()) === false) {
                return;
            }
        }
        await inst?.installOrUpdate();
        if (wasInstalled) {
            if ((0, AfterUpdate_js_1.implementsAfterUpdate)(inst)) {
                await inst.afterUpdate();
            }
        }
        else {
            if ((0, hooks_1.implementsAfterInstall)(inst)) {
                await inst.afterInstall();
            }
        }
        if ((0, AfterInstallOrUpdate_1.implementsAfterInstallOrUpdate)(inst)) {
            await inst.afterInstallOrUpdate();
        }
    }
    async uninstallById(idOrInstaller) {
        const inst = typeof idOrInstaller === "string" ? await this.createInstallerInstance(idOrInstaller) : idOrInstaller;
        const wasInstalled = (await inst?.isInstalled) ?? false;
        if (wasInstalled) {
            if ((0, BeforeUninstall_1.implementsBeforeUninstall)(inst)) {
                if ((await inst.beforeUninstall()) === false) {
                    return;
                }
            }
            await inst?.uninstall();
            if ((0, AfterUninstall_js_1.implementsAfterUninstall)(inst)) {
                await inst.afterUninstall();
            }
        }
    }
    async updateById(idOrInstaller) {
        const inst = typeof idOrInstaller === "string" ? await this.createInstallerInstance(idOrInstaller) : idOrInstaller;
        const wasInstalled = (await inst?.isInstalled) ?? false;
        if (wasInstalled) {
            if ((0, BeforeUpdate_1.implementsBeforeUpdate)(inst)) {
                if ((await inst.beforeUpdate()) === false) {
                    return;
                }
            }
            if ((0, BeforeInstallOrUpdate_1.implementsBeforeInstallOrUpdate)(inst)) {
                if ((await inst.beforeInstallOrUpdate()) === false) {
                    return;
                }
            }
            await inst?.update();
            if ((0, AfterUpdate_js_1.implementsAfterUpdate)(inst)) {
                await inst.afterUpdate();
            }
            if ((0, AfterInstallOrUpdate_1.implementsAfterInstallOrUpdate)(inst)) {
                await inst.afterInstallOrUpdate();
            }
        }
    }
    async getListItems(list) {
        return (await this.installers).join(list, (o) => o.id, (i) => i.id, (o, i) => ({
            ...i,
            installer: o,
        }));
    }
    async buildInstallerRunList(list) {
        const graph = new dependency_graph_1.DepGraph();
        for (const i of list) {
            await this.buildInstallerRunListAdd(graph, i.id);
        }
        const order = (0, dep_graph_1.depGraphToArray)(graph);
        return order;
    }
    async buildInstallerRunListAdd(graph, id) {
        const installerClass = await this.getPlatformInstallerById(id);
        if (!installerClass) {
            throw new Error(`Could not find installer with id "${id}"`);
        }
        const inst = new installerClass();
        graph.addNode(id, inst);
        if (inst.dependencies?.length) {
            for (const dep of inst.dependencies) {
                await this.buildInstallerRunListAdd(graph, dep);
                graph.addDependency(id, dep);
            }
        }
    }
    async update(list) {
        const installs = await this.buildInstallerRunList(list);
        for (const inst of installs) {
            await this.updateById(inst);
        }
    }
    async installOrUpdate(list) {
        const installs = await this.buildInstallerRunList(list);
        for (const inst of installs) {
            await this.installOrUpdateById(inst);
        }
    }
    async uninstall(list) {
        const installs = (await this.buildInstallerRunList(list)).reverse();
        for (const inst of installs) {
            await this.uninstallById(inst);
        }
    }
}
exports.InstallerService = InstallerService;
exports.GlobalInstallerService = new InstallerService();
