// @cpdevtools/lib-node-utilities@0.4.54/utils/cmd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.start = exports.run = exports.exec = exports.ChildProcessObservable = void 0;
const child_process_1 = require("child_process");
const rxjs_1 = require("rxjs");
class ChildProcessObservable extends rxjs_1.Observable {
    constructor(cmd, options) {
        super((subscriber) => {
            const sub = this._process$.subscribe(subscriber);
            return () => {
                sub.unsubscribe();
            };
        });
        Object.defineProperty(this, "cmd", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: cmd
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: options
        });
        Object.defineProperty(this, "_process$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new rxjs_1.Observable((subscriber) => {
                const child = (0, child_process_1.spawn)(this.cmd, this.options);
                child.once("error", (err) => {
                    subscriber.error(err);
                });
                child.once("close", (code) => {
                    subscriber.next({
                        type: "close",
                        code: code ?? 0,
                    });
                    subscriber.complete();
                });
                child.on("message", (message) => {
                    subscriber.next({
                        type: "message",
                        message,
                    });
                });
                child.stdout?.on("data", (data) => {
                    try {
                        const str = Buffer.from(data.filter((b) => !!b)).toString("ascii");
                        subscriber.next({
                            type: "data",
                            data: str,
                        });
                    }
                    catch (e) {
                        console.warn("Failed to parse process std output");
                        console.warn(e);
                    }
                });
                child.stderr?.on("data", (data) => {
                    try {
                        const str = Buffer.from(data.filter((b) => !!b)).toString("ascii");
                        subscriber.next({
                            type: "data",
                            data: str,
                        });
                    }
                    catch (e) {
                        console.warn("Failed to parse process std error output");
                        console.warn(e);
                    }
                });
                return () => {
                    child.unref();
                };
            }).pipe((0, rxjs_1.share)())
        });
        Object.defineProperty(this, "data$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this._process$.pipe((0, rxjs_1.filter)((d) => d.type === "data"))
        });
        Object.defineProperty(this, "dataComplete$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.data$.pipe((0, rxjs_1.reduce)((a, b) => {
                a.data += b.data;
                return a;
            }, {
                type: "data",
                data: "",
            }))
        });
    }
    get complete() {
        return (0, rxjs_1.lastValueFrom)(this._process$);
    }
}
exports.ChildProcessObservable = ChildProcessObservable;
async function exec(cmd, { cwd, env } = {}) {
    const child = new ChildProcessObservable(cmd, {
        shell: true,
        stdio: "inherit",
        cwd,
        env: env ?? process.env,
    });
    const result = await child.complete;
    return result.code;
}
exports.exec = exec;
async function run(cmd, { cwd, env } = {}) {
    const child = new ChildProcessObservable(cmd, {
        shell: true,
        stdio: "pipe",
        cwd,
        env: env ?? process.env,
    });
    const result = await (0, rxjs_1.lastValueFrom)(child.dataComplete$);
    const comp = await child.complete;
    if (comp.code !== 0) {
        throw {
            code: comp.code,
            data: result.data,
        };
    }
    return result.data;
}
exports.run = run;
async function start(cmd, { cwd } = {}) {
    const child = (0, child_process_1.spawn)(cmd, {
        shell: true,
        detached: true,
        stdio: "ignore",
        cwd,
        env: process.env,
    });
    child.unref();
}
exports.start = start;
