// @cpdevtools/lib-node-utilities@0.4.54/utils/docker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForDockerInit = exports.killDockerDesktop = exports.getDockerDesktopConfigPath = exports.throwIfDockerDesktopNotRunning = exports.throwIfDockerNotRunning = exports.isDockerRunning = exports.isDockerDesktopRunning = exports.restartDockerDesktop = exports.startDockerDesktop = exports.getDockerDesktopPath = exports.dockerLogin = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const chalk_1 = require("./chalk");
const cmd_1 = require("./cmd");
const sleep_1 = require("./sleep");
const windows_1 = require("./windows");
const wsl_1 = require("./wsl");
async function dockerLogin(url, user, token) {
    const chalk = await (0, chalk_1.importChalk)();
    console.info(`Attempting to log docker into ${chalk.blueBright(url)} with user ${chalk.yellowBright(user)}`);
    const result = await (0, cmd_1.exec)(`echo "${token}" | docker login ${url} -u ${user} --password-stdin`);
    return !result;
}
exports.dockerLogin = dockerLogin;
async function getDockerDesktopPath() {
    const path = await (0, wsl_1.translateWindowsPath)("C:\\Program Files\\Docker\\Docker");
    return path;
}
exports.getDockerDesktopPath = getDockerDesktopPath;
async function startDockerDesktop() {
    try {
        const cmd = `"${await getDockerDesktopPath()}/Docker Desktop.exe" &`;
        await (0, cmd_1.exec)(cmd);
        const dockerConfigPath = await getDockerDesktopConfigPath();
        while (!(0, fs_1.existsSync)(dockerConfigPath)) {
            await (0, sleep_1.sleep)(500);
        }
    }
    catch (e) {
        throw e;
    }
}
exports.startDockerDesktop = startDockerDesktop;
async function restartDockerDesktop() {
    const chalk = await (0, chalk_1.importChalk)();
    console.info(chalk.gray("Restarting Docker Desktop..."));
    await killDockerDesktop();
    await startDockerDesktop();
    await waitForDockerInit(true);
}
exports.restartDockerDesktop = restartDockerDesktop;
async function isDockerDesktopRunning() {
    return await (0, windows_1.isApplicationRunning)("Docker Desktop.exe");
}
exports.isDockerDesktopRunning = isDockerDesktopRunning;
async function isDockerRunning() {
    try {
        await throwIfDockerNotRunning();
        return true;
    }
    catch { }
    return false;
}
exports.isDockerRunning = isDockerRunning;
async function throwIfDockerNotRunning() {
    try {
        await (0, cmd_1.run)("docker info");
    }
    catch {
        throw new Error(`Docker is not running`);
    }
}
exports.throwIfDockerNotRunning = throwIfDockerNotRunning;
async function throwIfDockerDesktopNotRunning() {
    if (!(await isDockerDesktopRunning())) {
        throw new Error(`Docker Desktop is not running`);
    }
}
exports.throwIfDockerDesktopNotRunning = throwIfDockerDesktopNotRunning;
async function getDockerDesktopConfigPath() {
    const dataPath = (await (0, wsl_1.translateWindowsPath)(await (0, wsl_1.readWindowsEnv)("appdata"))).trim();
    return (0, path_1.join)(dataPath, "Docker", "settings.json");
}
exports.getDockerDesktopConfigPath = getDockerDesktopConfigPath;
async function killDockerDesktop() {
    try {
        return await (0, cmd_1.run)('taskkill.exe /IM "Docker Desktop.exe" /F');
    }
    catch {
        return false;
    }
}
exports.killDockerDesktop = killDockerDesktop;
async function waitForDockerInit(isRestart = false) {
    await (0, sleep_1.sleep)(100);
    const chalk = await (0, chalk_1.importChalk)();
    let c = 0;
    const headerDelay = isRestart ? 8 : 4;
    while (c !== -1) {
        try {
            if (c < headerDelay) {
                c++;
            }
            else if (c === headerDelay) {
                if (isRestart) {
                    console.info();
                    console.info(chalk.yellow("********************************************************************"));
                    console.info(chalk.yellow("* Waiting for docker to restart...                                 *"));
                    console.info(chalk.yellow("*                                                                  *"));
                    console.info(chalk.yellow("* If it fails to start automatically make sure it is running       *"));
                    console.info(chalk.yellow("* and if nessisarry execute a restart from the taskbar menu.       *"));
                    console.info(chalk.yellow("*                                                                  *"));
                    console.info(chalk.yellow("* Sure is taking it's time          (╯°□°)╯︵ ┻━┻                 *"));
                    console.info(chalk.yellow("********************************************************************"));
                    console.info();
                }
                else {
                    console.info();
                    console.info(chalk.yellow("********************************************************************"));
                    console.info(chalk.yellow("* Waiting for access to docker                                     *"));
                    console.info(chalk.yellow("*                                                                  *"));
                    console.info(chalk.yellow("* Please make sure that docker desktop is running and restart      *"));
                    console.info(chalk.yellow("* the service if nessisarry                                        *"));
                    console.info(chalk.yellow("********************************************************************"));
                    console.info();
                }
                c++;
            }
            await throwIfDockerNotRunning();
            if (c >= headerDelay) {
                if (isRestart) {
                    console.info(chalk.grey("Docker is ready.                     ┬─┬ノ(º_ºノ)"));
                }
                else {
                    console.info(chalk.grey("Docker is ready."));
                }
            }
            c = -1;
        }
        catch {
            await (0, sleep_1.sleep)(3750);
        }
    }
}
exports.waitForDockerInit = waitForDockerInit;
