// @cpdevtools/lib-node-utilities@0.4.54/utils/download.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileDownload = void 0;
const cli_progress_1 = __importDefault(require("cli-progress"));
const date_fns_1 = require("date-fns");
const fs_1 = __importDefault(require("fs"));
const rxjs_1 = require("rxjs");
const axios_1 = __importDefault(require("axios"));
function formatBytes(bytes, decimals = 2) {
    if (bytes === 0)
        return "0 Bytes";
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ["Bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + " " + sizes[i];
}
function toFormattedEvent(evt) {
    try {
        const eta = (0, date_fns_1.intervalToDuration)({ start: 0, end: evt.eta * 1000 });
        const duration = (0, date_fns_1.intervalToDuration)({ start: 0, end: evt.duration * 1000 });
        return {
            type: "progress",
            timeRemaining: (0, date_fns_1.formatDuration)(eta, {
                format: ["hours", "minutes", "seconds"],
            }),
            duration: (0, date_fns_1.formatDuration)(duration, {
                format: ["hours", "minutes", "seconds"],
            }),
            percent: `${Math.floor(evt.percent * 10000) / 100}%`,
            time: evt.time,
            size: formatBytes(evt.bytesTotal),
            received: formatBytes(evt.bytesReceived),
            rate: formatBytes(evt.bytesRate) + "/s",
        };
    }
    catch {
        return {
            type: "progress",
            timeRemaining: "",
            duration: "",
            percent: "",
            time: evt.time ?? new Date(),
            rate: "",
            received: "",
            size: "",
        };
    }
}
class FileDownload {
    constructor(sourceUrl, filePath) {
        Object.defineProperty(this, "sourceUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: sourceUrl
        });
        Object.defineProperty(this, "filePath", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: filePath
        });
        Object.defineProperty(this, "_downloadPromise", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_started", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_progressHistory", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "_events$$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new rxjs_1.Subject()
        });
        Object.defineProperty(this, "events$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this._events$$.asObservable()
        });
        Object.defineProperty(this, "progress$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.events$.pipe((0, rxjs_1.filter)((e) => e.type === "progress"))
        });
        Object.defineProperty(this, "progressFormatted$", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.progress$.pipe((0, rxjs_1.map)(toFormattedEvent))
        });
    }
    download(showProgressBar) {
        let bar = showProgressBar
            ? new cli_progress_1.default.SingleBar({
                format: "{bar} {percent} - {timeRemaining} @ {rate}",
                barsize: 40,
            })
            : null;
        if (!this._downloadPromise) {
            this._started = new Date();
            this._downloadPromise = new Promise(async (res, rej) => {
                this._events$$.next({
                    type: "connecting",
                    time: new Date(),
                });
                const response = await axios_1.default.request({
                    method: "get",
                    responseType: "stream",
                    url: this.sourceUrl,
                });
                const bytesTotal = +response.headers["content-length"] ?? -1;
                let bytesReceived = 0;
                bar?.start(bytesTotal, 0);
                this._events$$.next({
                    type: "connected",
                    time: new Date(),
                    contentType: response.headers["content-type"],
                    bytesTotal,
                });
                this._progressHistory.push([new Date(), 0]);
                const message = response.data;
                message.pipe(fs_1.default.createWriteStream(this.filePath));
                message.on("data", (data) => {
                    bytesReceived += data.length;
                    this._progressHistory.push([new Date(), bytesReceived]);
                    const stats5Sec = this.calcProgressStats(bytesReceived, bytesTotal, 5);
                    const stats30Sec = this.calcProgressStats(bytesReceived, bytesTotal, 30);
                    const event = {
                        ...stats30Sec,
                        bytesRate: stats5Sec.bytesRate,
                    };
                    this._events$$.next(event);
                    bar?.update(event.bytesReceived, toFormattedEvent(event));
                    this.pruneHistory();
                });
                message.on("end", () => {
                    bytesReceived = bytesTotal;
                    const stats5Sec = this.calcProgressStats(bytesReceived, bytesTotal, 5);
                    const stats30Sec = this.calcProgressStats(bytesReceived, bytesTotal, 30);
                    const event = {
                        ...stats30Sec,
                        bytesRate: stats5Sec.bytesRate,
                    };
                    this._events$$.next(event);
                    bar?.update(event.bytesReceived, toFormattedEvent(event));
                    bar?.stop();
                    const now = new Date();
                    this._events$$.next({
                        type: "complete",
                        time: now,
                        bytesTotal,
                        rate: stats5Sec.bytesRate,
                        duration: (0, date_fns_1.differenceInMilliseconds)(now, this._started) / 1000,
                    });
                    this._started = undefined;
                    this._progressHistory = [];
                    this._downloadPromise = undefined;
                    res();
                });
                message.on("error", (err) => {
                    this._events$$.next({
                        type: "error",
                        time: new Date(),
                        error: err,
                    });
                    this._started = undefined;
                    this._progressHistory = [];
                    this._downloadPromise = undefined;
                    bar?.stop();
                    rej(err);
                });
            });
        }
        return this._downloadPromise;
    }
    pruneHistory() {
        const now = new Date();
        this._progressHistory = this._progressHistory.filter((d) => (0, date_fns_1.differenceInSeconds)(now, d[0]) <= 30);
    }
    calcProgressStats(loaded, total, timeSpan = 5) {
        const now = new Date();
        const data = this._progressHistory.find((d) => (0, date_fns_1.differenceInSeconds)(now, d[0]) <= timeSpan);
        if (data) {
            const [sTime, sBytes] = data;
            const deltaTime = (0, date_fns_1.differenceInMilliseconds)(now, sTime);
            const deltaBytes = loaded - sBytes;
            const bytesRemaining = total - loaded;
            const rate = (deltaBytes / deltaTime) * 1000;
            const eta = bytesRemaining / rate;
            return {
                type: "progress",
                bytesRate: rate,
                bytesReceived: loaded,
                bytesTotal: total,
                eta: eta,
                percent: loaded / (total || -1),
                time: now,
                duration: (0, date_fns_1.differenceInMilliseconds)(now, this._started) / 1000,
            };
        }
        else {
            return {
                type: "progress",
                bytesRate: 0,
                bytesReceived: loaded,
                bytesTotal: total,
                eta: 0,
                percent: loaded / (total || -1),
                time: now,
                duration: (0, date_fns_1.differenceInMilliseconds)(now, this._started) / 1000,
            };
        }
    }
}
exports.FileDownload = FileDownload;
