// @cpdevtools/lib-node-utilities@0.4.54/utils/git.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gitHasChanges = exports.gitStatus = exports.gitSync = exports.gitAdd = exports.gitClone = exports.gitCommit = exports.gitPush = exports.gitPull = exports.gitIsRepo = exports.setConfig = exports.getConfig = void 0;
const fs_1 = require("fs");
const simple_git_1 = require("simple-git");
const cmd_1 = require("./cmd");
async function getConfig(name) {
    const git = (0, simple_git_1.simpleGit)();
    const result = await git.getConfig(name, "global");
    return result.value;
}
exports.getConfig = getConfig;
async function setConfig(name, value) {
    const git = (0, simple_git_1.simpleGit)(".");
    const result = await git.addConfig(name, value, false, "global");
    return result;
}
exports.setConfig = setConfig;
async function gitIsRepo(path) {
    if (!(0, fs_1.existsSync)(path)) {
        return false;
    }
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.checkIsRepo();
    return result;
}
exports.gitIsRepo = gitIsRepo;
async function gitPull(path) {
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.pull();
    return result;
}
exports.gitPull = gitPull;
async function gitPush(path) {
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.push();
    return result;
}
exports.gitPush = gitPush;
async function gitCommit(path, message) {
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.commit(message, undefined);
    return result;
}
exports.gitCommit = gitCommit;
async function gitClone(path, url) {
    await (0, cmd_1.exec)(`git clone ${url}`, { cwd: path });
}
exports.gitClone = gitClone;
async function gitAdd(path) {
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.add(".");
    return result;
}
exports.gitAdd = gitAdd;
async function gitSync(path) {
    if (await gitIsRepo(path)) {
        await gitAdd(path);
        await gitCommit(path, "Syncronizing");
        try {
            await gitPull(path);
        }
        catch { }
        await gitPush(path);
    }
}
exports.gitSync = gitSync;
async function gitStatus(path) {
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.status();
    return result;
}
exports.gitStatus = gitStatus;
async function gitHasChanges(path) {
    const git = (0, simple_git_1.simpleGit)(path);
    const result = await git.diffSummary();
    return result.files.length > 0;
}
exports.gitHasChanges = gitHasChanges;
