// @cpdevtools/lib-node-utilities@0.4.54/utils/github.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRepo = exports.cloneRepo = exports.configureGithubCli = exports.githubAuthStatus = exports.githubLogin = void 0;
const chalk_1 = require("./chalk");
const cmd_1 = require("./cmd");
async function githubLogin(...args) {
    const chalk = await (0, chalk_1.importChalk)();
    if (args.length === 2) {
        console.info(`Attempting to log into github.com with user ${chalk.yellowBright(args[0])}`);
        const result = await (0, cmd_1.exec)(`echo "${args[1]}" | gh auth login --with-token`);
        return !result;
    }
    const result = await (0, cmd_1.exec)(`GITHUB_TOKEN="${args[0]}"; gh auth login -h github.com`);
    return !result;
}
exports.githubLogin = githubLogin;
async function githubAuthStatus(env) {
    const result = await (0, cmd_1.run)(`gh auth status -t`, { env });
    const usernameRegExp = /Logged in to github\.com as ([\w\d]+)/i;
    const usernameMatch = result.match(usernameRegExp);
    const username = usernameMatch?.[1];
    const protocolRegExp = /configured to use ([\w]+) protocol/i;
    const protocolMatch = result.match(protocolRegExp);
    const protocol = protocolMatch?.[1];
    const tokenRegExp = /Token: ([\w\d\*]+)/i;
    const tokenMatch = result.match(tokenRegExp);
    const token = tokenMatch?.[1];
    const scopesRegExp = /Token scopes: (.*?)$/im;
    const scopesMatch = result.match(scopesRegExp);
    const scopes = scopesMatch?.[1]?.split(",").map((s) => s.trim());
    return {
        username,
        protocol,
        token,
        scopes,
    };
}
exports.githubAuthStatus = githubAuthStatus;
async function configureGithubCli() {
    console.info(`Configuring Github Cli`);
    await (0, cmd_1.exec)(`gh config set git_protocol https -h github.com`);
    await (0, cmd_1.exec)(`gh auth setup-git`);
}
exports.configureGithubCli = configureGithubCli;
async function cloneRepo(id, path, opts) {
    await (0, cmd_1.exec)(`gh repo clone ${id} ${path ?? ""}`, { cwd: opts?.cwd });
}
exports.cloneRepo = cloneRepo;
async function createRepo(id, opts) {
    const cmd = [`gh repo create ${id}`];
    if (opts?.clone) {
        cmd.push(`--clone`);
    }
    if (opts?.description) {
        cmd.push(`--description "${opts?.description}"`);
    }
    if (opts?.disableIssues) {
        cmd.push(`--disable-issues`);
    }
    if (opts?.disableWiki) {
        cmd.push(`--disable-wiki`);
    }
    if (opts?.gitIgnore) {
        cmd.push(`--gitignore "${opts?.gitIgnore}"`);
    }
    if (opts?.homepage) {
        cmd.push(`--homepage "${opts?.homepage}"`);
    }
    if (opts?.includeAllBranches) {
        cmd.push(`--include-all-branches`);
    }
    if (opts?.internal) {
        cmd.push(`--internal`);
    }
    if (opts?.license) {
        cmd.push(`--license "${opts?.license}"`);
    }
    if (opts?.private) {
        cmd.push(`--private`);
    }
    if (opts?.public) {
        cmd.push(`--public`);
    }
    if (opts?.push) {
        cmd.push(`--push`);
    }
    if (opts?.remote) {
        cmd.push(`--remote "${opts?.remote}"`);
    }
    if (opts?.source) {
        cmd.push(`--source "${opts?.source}"`);
    }
    if (opts?.team) {
        cmd.push(`--team "${opts?.team}"`);
    }
    if (opts?.template) {
        cmd.push(`--template "${opts?.template}"`);
    }
    await (0, cmd_1.exec)(cmd.join(" "), { cwd: opts?.cwd });
}
exports.createRepo = createRepo;
/*
export async function searchRepos() {
  const a = new Octokit({auth: process.env.GITHUB_TOKEN});

  return await a.repos.listForAuthenticatedUser({type: 'all', });
}
*/
