// @cpdevtools/lib-node-utilities@0.4.54/utils/vscode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.killVsCode = exports.uninstallVSCodeExtension = exports.installVSCodeExtension = exports.launchVSCodeDevContainer = exports.launchContainerUrl = exports.getContainerLaunchUrl = exports.launchVSCode = void 0;
const path_1 = require("path");
const child_process_1 = require("child_process");
const posix_1 = __importDefault(require("path/posix"));
const cmd_1 = require("./cmd");
const wsl_1 = require("./wsl");
const is_wsl_1 = __importDefault(require("is-wsl"));
function launchVSCode(path = ".") {
    (0, child_process_1.spawn)(`code ${path}`, { shell: true, detached: true, stdio: "ignore" });
}
exports.launchVSCode = launchVSCode;
async function getContainerLaunchUrl(containerPath, workspace) {
    const hexPath = Buffer.from(await (0, wsl_1.translateWslPath)(containerPath)).toString("hex");
    return `vscode-remote://dev-container+${posix_1.default.join(hexPath, workspace ?? "")}`;
}
exports.getContainerLaunchUrl = getContainerLaunchUrl;
function launchContainerUrl(launchUrl) {
    const isWS = launchUrl.endsWith(".code-workspace");
    const flag = isWS ? "file-uri" : "folder-uri";
    const cmd = `code --${flag} "${launchUrl}"`;
    (0, child_process_1.spawn)(cmd, { shell: true, detached: true, stdio: "ignore" });
}
exports.launchContainerUrl = launchContainerUrl;
async function launchVSCodeDevContainer(containerPath = ".", open) {
    const isWS = (0, path_1.extname)(open ?? "") === ".code-workspace";
    const flag = isWS ? "file-uri" : "folder-uri";
    const hexPath = Buffer.from(await (0, wsl_1.translateWslPath)(containerPath)).toString("hex");
    let uri = `vscode-remote://dev-container+${hexPath}/${open ?? ""}`;
    const cmd = `code --${flag} "${uri}"`;
    (0, child_process_1.spawn)(cmd, { shell: true, detached: true, stdio: "ignore" });
}
exports.launchVSCodeDevContainer = launchVSCodeDevContainer;
async function installVSCodeExtension(idOrPath, options) {
    const cmd = is_wsl_1.default ? "cd /mnt/c/ && cmd.exe /c code" : "code";
    const command = `${cmd} --install-extension ${idOrPath} ${options?.preRelease ? "--pre-release" : ""} ${options?.force ? "--force" : ""}`;
    await (0, cmd_1.exec)(command);
}
exports.installVSCodeExtension = installVSCodeExtension;
async function uninstallVSCodeExtension(id) {
    const command = `code --uninstall-extension ${id} `;
    await (0, cmd_1.exec)(command);
}
exports.uninstallVSCodeExtension = uninstallVSCodeExtension;
async function killVsCode() {
    try {
        return await (0, cmd_1.run)('taskkill.exe /IM "Code.exe" /F');
    }
    catch {
        return false;
    }
}
exports.killVsCode = killVsCode;
