// @cpdevtools/lib-node-utilities@0.4.54/utils/windows.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runAsWindowsAdmin = exports.execAsWindowsAdmin = exports.isApplicationRunning = exports.removeRunOnceAfterRestart = exports.runOnceAfterRestart = exports.isWin11 = exports.isWin10 = exports.windowsVersion = exports.windowsVer = exports.isWindows = exports.locateInstallationPath = exports.loadAppxPackage = exports.loadAppxPackages = void 0;
const os_1 = __importDefault(require("os"));
const semver_1 = __importDefault(require("semver"));
const cmd_1 = require("./cmd");
const wsl_1 = require("./wsl");
async function loadAppxPackages() {
    const result = await (0, cmd_1.run)(`powershell.exe Get-AppxPackage"`);
    return result
        .split("\n\n")
        .map((appStr) => appStr.trim().split("\n"))
        .map((lines) => {
        const result = {};
        lines.forEach((line) => {
            const p = line.split(":", 2);
            result[p[0].trim()] = p[1].trim();
        });
        return result;
    });
}
exports.loadAppxPackages = loadAppxPackages;
async function loadAppxPackage(appxPackageName) {
    const packages = await loadAppxPackages();
    return packages.find((p) => p.Name === "appxPackageName");
}
exports.loadAppxPackage = loadAppxPackage;
async function locateInstallationPath(appxPackageName) {
    const pkg = await loadAppxPackage(appxPackageName);
    return pkg?.InstallLocation ? (0, wsl_1.translateWindowsPath)(pkg?.InstallLocation) : undefined;
}
exports.locateInstallationPath = locateInstallationPath;
exports.isWindows = os_1.default.platform() === "win32";
exports.windowsVer = exports.isWindows ? os_1.default.release() : null;
exports.windowsVersion = exports.isWindows ? semver_1.default.parse(exports.windowsVer) : null;
exports.isWin10 = !exports.isWindows ? false : exports.windowsVersion?.major === 10 && exports.windowsVersion?.patch !== 22000;
exports.isWin11 = !exports.isWindows ? false : exports.windowsVersion?.major === 10 && exports.windowsVersion?.patch === 22000;
async function runOnceAfterRestart(id, script) {
    await (0, cmd_1.run)(`reg.exe add "HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\RunOnce" /v "${id}" /d "${script}" /f`);
}
exports.runOnceAfterRestart = runOnceAfterRestart;
async function removeRunOnceAfterRestart(id) {
    await (0, cmd_1.run)(`reg.exe delete "HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\RunOnce" /v "${id}" /f`);
}
exports.removeRunOnceAfterRestart = removeRunOnceAfterRestart;
async function isApplicationRunning(name) {
    try {
        await (0, cmd_1.run)(`tasklist.exe /fi "ImageName eq ${name}" | find /I "${name}"`);
        return true;
    }
    catch { }
    return false;
}
exports.isApplicationRunning = isApplicationRunning;
async function execAsWindowsAdmin(cmd, opts = {}) {
    return await (0, cmd_1.exec)(`powershell.exe Start-Process cmd.exe -Wait -Verb runAs -ArgumentList @('/c', '${cmd.join(`', '`)}')`, opts);
}
exports.execAsWindowsAdmin = execAsWindowsAdmin;
async function runAsWindowsAdmin(cmd, opts = {}) {
    return await (0, cmd_1.run)(`powershell.exe Start-Process cmd.exe -Wait -Verb runAs -ArgumentList @('/c', '${cmd.join(`', '`)}')`, opts);
}
exports.runAsWindowsAdmin = runAsWindowsAdmin;
