// @cpdevtools/lib-node-utilities@0.4.54/utils/winget.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateOrInstallWinget = exports.updateWinget = exports.uninstallWinget = exports.installWinget = exports.isValidInstallFile = exports.wingetInfo = exports.isInstalledWsl = void 0;
const fs_1 = require("fs");
const sha256_file_1 = __importDefault(require("sha256-file"));
const yaml_1 = __importDefault(require("yaml"));
const chalk_1 = require("./chalk");
const cmd_1 = require("./cmd");
async function isInstalledWsl(id) {
    try {
        await (0, cmd_1.run)(`winget.exe list -e --id ${id}`);
        return true;
    }
    catch (e) {
        if (e.code === 20) {
            return false;
        }
        throw e;
    }
}
exports.isInstalledWsl = isInstalledWsl;
async function wingetInfo(id) {
    let result = await (0, cmd_1.run)(`winget.exe show -e --id ${id}`);
    result = result.slice(result.indexOf("Version:"));
    const raw = yaml_1.default.parse(result);
    return {
        id,
        author: raw.Author,
        copyright: raw.Copyright,
        description: raw.Description,
        homepage: raw.Homepage,
        license: raw.License,
        moniker: raw.Moniker,
        publisher: raw.Publisher,
        publisherSupportUrl: raw["Publisher Support Url"],
        publisherUrl: raw["Publisher Url"],
        version: raw.Version,
        installer: {
            downloadUrl: raw.Installer["Download Url"],
            locale: raw.Installer.Locale,
            releaseDate: raw.Installer["Release Date"],
            SHA256: raw.Installer.SHA256,
            type: raw.Installer.Type,
        },
    };
}
exports.wingetInfo = wingetInfo;
async function isValidInstallFile(filepath, id) {
    if ((0, fs_1.existsSync)(filepath)) {
        const [info, sha] = await Promise.all([
            wingetInfo(id),
            new Promise((res, rej) => (0, sha256_file_1.default)(filepath, (error, check) => (error ? rej(error) : res(check)))),
        ]);
        return info.installer.SHA256 === sha;
    }
    return false;
}
exports.isValidInstallFile = isValidInstallFile;
async function installWinget(id) {
    const chalk = await (0, chalk_1.importChalk)();
    console.info(chalk.greenBright(`${chalk.cyanBright(id)}: Installing...`));
    console.info();
    const result = await (0, cmd_1.exec)(`winget.exe install -e --id ${id}`);
    console.info();
    console.info();
    return result;
}
exports.installWinget = installWinget;
async function uninstallWinget(id) {
    const chalk = await (0, chalk_1.importChalk)();
    console.info(chalk.greenBright(`${chalk.cyanBright(id)}: Uninstalling...`));
    console.info();
    const result = await (0, cmd_1.exec)(`winget.exe uninstall -e --id ${id}`);
    console.info();
    console.info();
    return result;
}
exports.uninstallWinget = uninstallWinget;
async function updateWinget(id, args) {
    const chalk = await (0, chalk_1.importChalk)();
    console.info(chalk.blueBright(`${chalk.cyanBright(id)}: Checking for updates...`));
    console.info();
    const result = await (0, cmd_1.exec)(`winget.exe upgrade -he --verbose-logs --id ${id}${args ? " " + args : ""}`);
    console.info();
    console.info();
    return result;
}
exports.updateWinget = updateWinget;
async function updateOrInstallWinget(id) {
    if (await isInstalledWsl(id)) {
        return await updateWinget(id);
    }
    return await installWinget(id);
}
exports.updateOrInstallWinget = updateOrInstallWinget;
