// @cpdevtools/lib-node-utilities@0.4.54/utils/wsl.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readWindowsEnv = exports.listWslDistributions = exports.isWslDistroInstalled = exports.updateWSL = exports.installWSLKernelUpdate = exports.installWSL = exports.isWslInstalled = exports.getWslDefaultVersion = exports.getWslVersion = exports.translateWslPath = exports.translateWindowsPath = void 0;
const promises_1 = require("fs/promises");
const path_1 = __importDefault(require("path"));
const semver_1 = __importDefault(require("semver"));
const chalk_1 = require("./chalk");
const cmd_1 = require("./cmd");
const windows_1 = require("./windows");
async function translateWindowsPath(path) {
    const result = (await (0, cmd_1.run)(`wslpath -a -u "${path}"`)).trim();
    return result;
}
exports.translateWindowsPath = translateWindowsPath;
async function translateWslPath(path) {
    try {
        return (await (0, cmd_1.run)(`wslpath -a -w "${path}"`)).trim();
    }
    catch {
        return path;
    }
}
exports.translateWslPath = translateWslPath;
async function getWslVersion() {
    try {
        const result = ((await (0, cmd_1.run)(`wsl.exe --status`)) ?? "").split("\n").map((l) => l.trim());
        const kernelError = !!result.find((l) => l.includes("kernel file is not found"));
        if (kernelError) {
            return semver_1.default.parse("1.0.0");
        }
        const verStr = result
            .find((l) => l.includes("Kernel version:"))
            ?.split(":")?.[1]
            ?.trim();
        if (verStr) {
            const p = verStr.split(".").slice(0, 3);
            return semver_1.default.parse(`${p.join(".")}`, true);
        }
    }
    catch { }
    return null;
}
exports.getWslVersion = getWslVersion;
async function getWslDefaultVersion() {
    try {
        const verStr = (await (0, cmd_1.run)(`wsl.exe --status`))
            .split("\n")
            .map((l) => l.trim())
            .find((l) => l.includes("Default Version:"))
            ?.split(":")?.[1]
            ?.trim() ?? "0";
        return Number.isNaN(+verStr) ? 0 : +verStr;
    }
    catch { }
    return 0;
}
exports.getWslDefaultVersion = getWslDefaultVersion;
async function isWslInstalled() {
    return (await getWslVersion()) !== null;
}
exports.isWslInstalled = isWslInstalled;
async function installWSL() {
    if (windows_1.isWindows) {
        const p = path_1.default.join(process.env["temp"] ?? "", `installWSL.cmd`);
        await (0, promises_1.writeFile)(p, `
dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart
dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart
    `, { encoding: "utf-8" });
        await (0, windows_1.execAsWindowsAdmin)([p]);
        return true;
    }
    return false;
}
exports.installWSL = installWSL;
async function installWSLKernelUpdate() {
    const p = path_1.default.join(process.env["temp"] ?? "", `installWSLKernelUpdate.cmd`);
    await (0, promises_1.writeFile)(p, `
curl --ssl https://wslstorestorage.blob.core.windows.net/wslblob/wsl_update_x64.msi > uwsl.msi && uwsl.msi
wsl --update
wsl --shutdown
wsl --set-default-version 2
`, { encoding: "utf-8" });
    await (0, windows_1.execAsWindowsAdmin)([p]);
}
exports.installWSLKernelUpdate = installWSLKernelUpdate;
async function updateWSL() {
    const chalk = await (0, chalk_1.importChalk)();
    if (windows_1.isWindows) {
        console.info(chalk.gray(`Updating wsl...`));
        await (0, windows_1.execAsWindowsAdmin)([`wsl`, `--update`]);
        console.info(chalk.gray(`Restarting wsl...`));
        await (0, windows_1.execAsWindowsAdmin)([`wsl`, `--shutdown`]);
    }
}
exports.updateWSL = updateWSL;
async function isWslDistroInstalled(name) {
    return (await listWslDistributions()).includes(name);
}
exports.isWslDistroInstalled = isWslDistroInstalled;
async function listWslDistributions() {
    try {
        return (await (0, cmd_1.run)("wsl.exe --list"))
            .split("\n")
            .slice(1)
            .map((l) => {
            let r = l.trim();
            if (r.toLowerCase().endsWith(" (default)")) {
                r = r.slice(0, -10).trim();
            }
            return r;
        })
            .filter((l) => !!l);
    }
    catch { }
    return [];
}
exports.listWslDistributions = listWslDistributions;
async function readWindowsEnv(name) {
    return cleanValue(await (0, cmd_1.run)(`cmd.exe /c echo %${name}%`)).trim();
}
exports.readWindowsEnv = readWindowsEnv;
function cleanValue(val) {
    val = val.trim().split("\n").pop()?.trim() ?? "";
    if ((val.startsWith(`'`) && val.endsWith(`'`)) || (val.startsWith(`"`) && val.endsWith(`"`))) {
        return val.slice(1, -1);
    }
    return val;
}
